### PR TITLE
Feature/mtx writer

### PR DESCRIPTION
**Pull request by denizumuteser**
Whats done:
-mtx_writer is added.
-mtx_writer_tests are added.
-added exception handling to mtx_writer.
-added WriteCOOToMTX, WriteCSRToMTX, WriteArrayToMTX.
-added to IOBase.

### CODE REVIEW

File: tutorials/my_test/test.cc
Comment (AmroAlJundi) [inline tutorials/my_test/test.cc:1]: Why was this file added to tutorials? If it's just an example code snippet then it should go in the "examples" directory.

File: src/sparsebase/io/mtx_writer.cc [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+      symmetry_(symmetry) {}
+
+template <typename IDType, typename NNZType, typename ValueType>
+void MTXWriter<IDType, NNZType, ValueType>::WriteCOO(
+    format::COO<IDType, NNZType, ValueType> *coo) const {
+      
+      //illegal parameter checks
+      if (object_ != "matrix" && object_ != "vector") 
+      {
+        throw utils::ReaderException("Illegal value for the 'object' option in matrix market header");
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:29]: These exceptions should be to a "WriterException" rather than a "ReaderException". Could you create a new exception class for writer exceptions and use it instead for exceptions in Writer classes?

File: src/sparsebase/io/mtx_writer.cc [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+      
+      //header
+      mtxFile << "%%MatrixMarket " << object_ << " " << format_ << " " << field_ << " " << symmetry_ << "\n";
+      
+
+      //TODO add warning when given field and actual data format is not same (integer vs float etc)
+
+      if constexpr (std::is_same_v<ValueType, void>)
+      {
+        throw utils::WriterException("Cannot write an MTX with void ValueType");
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:71]: Actually, this can be done. It is done when the field is "pattern".

File: src/sparsebase/io/mtx_writer.cc [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+      else
+      {
+        auto dimensions = coo->get_dimensions();
+        IDType* row = coo->get_row();
+        IDType* col = coo->get_col();
+        ValueType* val = coo->get_vals();
+
+        if (val != NULL && field_ == "pattern")
+        {
+          std::cerr << "Warning: Pattern selected but values given.\n";
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:82]: You should use the logger here instead of std::cerr.

File: src/sparsebase/io/mtx_writer.cc [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+                      }
+                        
+                    }
+                    if (!found_symmetric) {
+                        throw utils::ReaderException("Matrix is not symmetric!");
+                    }
+                }
+                else //diagonal
+                {
+                  count_diagonal++;
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:119]: Here, if the symmetry is skew-symmetric, then there is an exception should be thrown. This is because diagonal values should be 0 for skew-symmetric matrices.

File: src/sparsebase/io/mtx_writer.cc [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+          }
+          std::sort(sort_vec.begin(), sort_vec.end(),
+                    [](std::tuple<IDType, IDType, ValueType> t1, std::tuple<IDType, IDType, ValueType> t2) {
+                      if (std::get<0>(t1) == std::get<0>(t2)) {
+                        return std::get<1>(t1) < std::get<1>(t2);
+                      }
+                      return std::get<0>(t1) < std::get<0>(t2);
+                    });
+
+          for (int i = 0; i < coo->get_num_nnz(); i++) {
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:160]: This loop doesn't seem to be doing anything, I guess it was missed :)

**Review by AmroAlJundi — CHANGES_REQUESTED** [review]
Very well-written at well-tested.

I've left some comments that need to be addressed. Once addressed, it would be ready for merging.

File: tutorials/my_test/test.cc
Comment (denizumuteser) [reply tutorials/my_test/test.cc:1]: It was accidently added, not related, deleting.

File: src/sparsebase/io/mtx_writer.cc (hunk 1 above)
Comment (denizumuteser) [reply src/sparsebase/io/mtx_writer.cc:29]: Turns out, there is already a ReaderException class :), fixed now.

File: src/sparsebase/io/mtx_writer.cc (hunk 3 above)
Comment (denizumuteser) [reply src/sparsebase/io/mtx_writer.cc:82]: fixed.

File: src/sparsebase/io/mtx_writer.cc (hunk 4 above)
Comment (denizumuteser) [reply src/sparsebase/io/mtx_writer.cc:119]: I added a check for diagonal entries if any of them is non zero and symmetry is skew symmetric, I throw an exception now.

File: src/sparsebase/io/mtx_writer.cc (hunk 5 above)
Comment (denizumuteser) [reply src/sparsebase/io/mtx_writer.cc:160]: deleted :)

File: src/sparsebase/io/mtx_writer.cc (hunk 2 above)
Comment (denizumuteser) [reply src/sparsebase/io/mtx_writer.cc]: that is true, hopefully resolved.

File: src/sparsebase/io/mtx_writer.cc [hunk 6]
```diff
@@ -63,87 +64,135 @@ void MTXWriter<IDType, NNZType, ValueType>::WriteCOO(
       //header
       mtxFile << "%%MatrixMarket " << object_ << " " << format_ << " " << field_ << " " << symmetry_ << "\n";
       
-
       //TODO add warning when given field and actual data format is not same (integer vs float etc)
 
       if constexpr (std::is_same_v<ValueType, void>)
       {
-        throw utils::WriterException("Cannot write an MTX with void ValueType");
+        if (field_ == "pattern")
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:71]: I think this should be != right?

File: src/sparsebase/io/mtx_writer.cc [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+      }
+
+      //Symmetry check
+      bool saidSymmetric = (symmetry_ == "symmetric" || symmetry_ == "skew-symmetric" || symmetry_ == "hermitian");
+      int NNZ = coo->get_num_nnz();
+      int count_symmetric = 0;
+      int count_diagonal = 0;
+      int is_diagonal_all_zero = true;
+      if (saidSymmetric && dimensions[0] == dimensions[1]) 
+      {
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:93]: What if `saidSymmetric == True` but `dimensions[0]!=dimensions[1]`. I think in that case, an exception should be thrown as well.

File: src/sparsebase/io/mtx_writer.cc [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+          {
+            if (!is_diagonal_all_zero)
+            {
+              throw utils::WriterException("Skew-symmetric matrix with non-zero diagonal values!");
+            }
+            NNZ -= (count_symmetric/2) + count_diagonal;
+          }
+          else
+          {
+            NNZ -= (count_symmetric/2);
```
Comment (AmroAlJundi) [inline src/sparsebase/io/mtx_writer.cc:169]: I think this should include the `+ count_diagonal` term, right? Since `symmetric` will have an element for each non-diagonal pair + an element for each element on the diagonal.

**Review by AmroAlJundi — CHANGES_REQUESTED** [review]
Check my comments.

**Review by AmroAlJundi — DISMISSED** [review]
Looks great!

**Review by SinanEkm — APPROVED** [review]
